### ISSUE TITLE
Feature flag use of redis grows unbounded

### ISSUE THREAD

**Issue by keegancsmith** [sourcegraph/sourcegraph#47887]
Looked deeply into the featureflag stuff when I did the redis conversion. It has some issues I am documenting here.

-   We won't delete a feature flag unless a some deletes in admin view. But feature flags can and will be evaled all the time in code.
-   We never expire, so if we just once eval a feature flag for a user it will live forever. This is concerning for anon users, those hsets will grow huge.
-   It doesn't need to be in store, it can just be cache.

I think the motivation for storing in redis was we wanted to log feature flag stuff across requests? Hard to understand to be honest, because we have things like LogBackendEvent which seems like it should just be the feature flags used in that backend request. Maybe user specific events can include everything? Either way it seems something is funky here.

I also think Cache is likely more approriate than Store, especially with the downsides mentioned of no expiry. Right now for sourcegraph.com this is taking up 184MB of memory in the redis store. See details below.

[Feature Flag PR](https://github.com/sourcegraph/sourcegraph/pull/35359)

### Size of hsets in prod

Investigated this <span class="timestamp-wrapper"><span class="timestamp">[2023-02-20 Mon]</span></span>.

    kubectl port-forward svc/redis-store 6379

```shell
redis-cli KEYS 'ff_*' | while read ff; do
    echo -n "$ff ";
    redis-cli debug object $ff | awk 'BEGIN { FS=":"; RS=" " } $1 == "serializedlength" { print $2 }'
done
```

| Redis Key                                 | serializedlength |
|----------------------------------------------|------------------|
| ff\_admin-analytics-enabled                  | 1783             |
| ff\_admin-analytics-cache-disabled           | 825              |
| ff\_search-ranking                           | 11113861         |
| ff\_cc-repo-aware-monitors                   | 1273             |
| ff\_search-hybrid                            | 19879531         |
| ff\_enable-streaming-git-blame               | 1034739          |
| ff\_search-aggregation-filters               | 254510           |
| ff\_ab-lucky-search                          | 30               |
| ff\_ab-visitor-tour-with-notebooks           | 43062359         |
| ff\_search-debug                             | 10623513         |
| ff\_cody                                     | 1582407          |
| ff\_new-repo-page                            | 8623625          |
| ff\_enable-embed-route                       | 21567            |
| ff\_quick-start-tour-for-authenticated-users | 168800           |
| ff\_contrast-compliant-syntax-highlighting   | 55272119         |
| ff\_accessible-file-tree                     | 2768555          |
| ff\_ab-email-verification-alert              | 14616            |
| ff\_search-content-based-lang-detection      | 20440027         |
| ff\_grpc                                     | 5142939          |
| ff\_search-input-show-history                | 3890392          |

So that is 183897471 bytes, or 184MB.

**Comment by keegancsmith** [comment]
cc @erzhtor @mrnugget @camdencheek this was stuff I wrote down when I did my redis stuff, finally filed an issue / investigated the impact on prod.

I don't want this issue to linger and never get resolved, so any ideas on how to resolve this? I don't think it should be a huge engineering effort.

**Comment by thenamankumar** [comment]
So the reason we implemented a cache to store evaluated feature flag was to support the AB testing usecase of lazy ff evaluation and automatically pass evaluated feature flag variants to all event logs. It is only used for AB testing purposes, but these days, we rarely perform any AB tests.

1. 184mb memory on our biggest instance ever? Does not sound too bad to me.  But yes, we can improve on this. 
2. @philipp-spiess recently discussed implementing “preloaded feature flags”, which can be made default and then lazy evaluation and caching can be made limited to the feature flags used for ab tests. 
3. Engineers should remove the feature flags when not in use. 

I'll follow up with @philipp-spiess and discuss how we can tackle this along with making improvements on the frontend as well.

**Comment by erzhtor** [comment]
Thanks @keegancsmith, for raising issue.

> It is only used for AB testing purposes, but these days, we rarely perform any AB tests.

This ^. @thenamankumar, however, I would suggest to involve product team on this, whether we want to support AB tests or are we gonna just use feature flags for partial/gradual rollout purposes only.

> @philipp-spiess recently discussed implementing “preloaded feature flags”, which can be made default and then lazy evaluation and caching can be made limited to the feature flags used for ab tests.

@thenamankumar, there is concern from @valerybugakov related to this, specifically, that feature flags are evaluated on the backend, thus blocking some rendering. Maybe we should group all work items related to feature flags into pr-faq or smth.

**Comment by keegancsmith** [comment]
> So the reason we implemented a cache to store evaluated feature flag was to support the AB testing usecase of lazy ff evaluation and automatically pass evaluated feature flag variants to all event logs. It is only used for AB testing purposes, but these days, we rarely perform any AB tests.

When I tried to wrap my head around this, what I couldn't understand is why you store the evaluation for ever, rather than just for that request.

> 1. 184mb memory on our biggest instance ever? Does not sound too bad to me.  But yes, we can improve on this.

The total memory use of the whole redis-store is 350MB. So over half of redis-store is storing data for feature flags which will never be cleaned up or read again.

> 2. @philipp-spiess recently discussed implementing “preloaded feature flags”, which can be made default and then lazy evaluation and caching can be made limited to the feature flags used for ab tests.

Sounds good. I still need more context on why we do things across requests rather than calculating stuff in a request.

> 3. Engineers should remove the feature flags when not in use.

The feature flag will still not be cleaned up in redis. The admin needs (for every instance) needs to explicitly both create the feature flag then hit delete in the admin UI. Note: we write to redis for every feature flag present in our codebase, not every feature flag an admin has created in the UI.

> @thenamankumar, there is concern from @valerybugakov related to this, specifically, that feature flags are evaluated on the backend, thus blocking some rendering. Maybe we should group all work items related to feature flags into pr-faq or smth.

What are the details here? feature flag evaluation should be very fast, have we seen cases where this is not the case?

**Comment by valerybugakov** [comment]
> What are the details here? feature flag evaluation should be very fast, have we seen cases where this is not the case?

Thanks for tagging me here, @erzhtor! Details: the feature flag evaluation itself is fast. The problem is that we send separate API requests over the network to evaluate feature flags, and it introduces additional client-side render blocks and causes redundant layout shits because we are forced to show empty space or loaders in the place where we might need to render content depending on the feature flag value for the current user.

What I plan to do as a part of the Storm project: preload all feature flags required for a page (if we know for sure that any user will need to feature flag up front), and there seem to be no issues with that. @erzhtor, pointed out that we can improve the solution by moving the feature flag evaluation to the client. I don't have much context on implementation details and trade-offs and would love to learn more about this if anybody already gave it a thought.

**Comment by camdencheek** [comment]
Just to add some color here: feature flags were originally designed to be not store any per-user state except user-specific overrides. When we were doing A/B tests, one common problem was that we stored the user's evaluated feature flags with events, but we did not have any way to determine whether a flag was actually _used_. This meant all our users were considered part of every A/B test and there was no way to filter to "just the users that actually saw the effects of the flag." My understanding is that this is why redis was added to the flow: to store which flags a user _actually_ saw. 

I think the reason we store in redis rather than just holding on to it per request is that sometimes the visible effect of the flag is in a different request than the measurable outcome. For example, a change to the search UI: a flag changes the experience of typing in a search query, but the outcome we want to measure is "number of searches," which is logged in events in the backend.

**Comment by keegancsmith** [comment]
> The problem is that we send separate API requests over the network to evaluate feature flags.

That indeed sounds sub-optimal. Given the design of the system though, we don't know upfront what the feature flags are since they are lazy. Sounds like we need to make them less lazy so we can frontload the work more. It's not as if the number of active feature flags in the codebase will ever get more than a few hundred anytime soon (given my analysis above we have only ever had 19 so far ever eval on dot-com)

> I think the reason we store in redis rather than just holding on to it per request is that sometimes the visible effect of the flag is in a different request than the measurable outcome. For example, a change to the search UI: a flag changes the experience of typing in a search query, but the outcome we want to measure is "number of searches," which is logged in events in the backend.

Thank you for that context. So if I understand the issue is we don't know the set of feature flags that could affect a user, so we need to store _somewhere_ the feature flags that have been evalled for a user. We don't actually need to know the values, since we could just ask the featureflag system again what the current value is?

To me it sounds like what we should be doing is using a redis set per user of which flags are evalled for it. Then you can set a conservative expire on that key.

**Comment by camdencheek** [comment]
> We don't actually need to know the values, since we could just ask the featureflag system again what the current value is?

Yep, that's right. The only caveat is if the evaluation rules change for a user (e.g. a new override or rollout % changes) the value of the flag might change. I think this would be a problem in the existing system too though.

> To me it sounds like what we should be doing is using a redis set per user of which flags are evalled for it. Then you can set a conservative expire on that key.

That sounds reasonable to me 👍

**Comment by philipp-spiess** [comment]
When I look at the feature flags currently active on dotcom, I wonder if we can't do a much easier solution that will give us almost all of the benefits and maybe won't require many changes?

From the 26 feature flags we currently have active on dotcom, only 5 are actually in an experiment with a partial rollout. My idea is that all of the feature flags that are either set via boolean values or that have partial rollout values of 0 or 100%, could be bypassing all of the lazy evaluation logic and be seen as static instead. On the frontend, if we would be able to have a GraphQL API that would just give us all feature flags that are static that we can preload on the initial render, this would already help us dramatically.

From the table that @keegancsmith posted above, only 3 feature flags would qualify as being evaluated this way which would total to a bit over 40MB (as opposed to the 184MB).

(This is kind-of similar to @thenamankumar [proposal written in Slack](https://sourcegraph.slack.com/archives/C04KDBTLG5P/p1676918067690919?thread_ts=1676916257.251059&cid=C04KDBTLG5P) but without the need to add a custom flag.
